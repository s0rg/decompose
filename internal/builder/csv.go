package builder

import (
	"encoding/csv"
	"fmt"
	"io"
	"strings"

	"github.com/s0rg/decompose/internal/node"
)

var csvHeader = []string{
	"service", "listen", "outbounds",
}

type CSV struct {
	j *JSON
}

func NewCSV() *CSV {
	return &CSV{
		j: NewJSON(),
	}
}

func (c *CSV) Name() string {
	return "csv"
}

func (c *CSV) AddNode(n *node.Node) error {
	return c.j.AddNode(n)
}

func (c *CSV) AddEdge(srcID, dstID string, port *node.Port) {
	c.j.AddEdge(srcID, dstID, port)
}

func (c *CSV) Write(w io.Writer) error {
	cw := csv.NewWriter(w)
	cw.UseCRLF = true

	_ = cw.Write(csvHeader)

	c.j.Sorted(func(n *node.JSON, _ bool) {
		_ = cw.Write([]string{
			n.Name,
			strings.Join(n.Listen, "\r\n"),
			renderOutbounds(n.Connected),
		})
	})

	cw.Flush()

	if err := cw.Error(); err != nil {
		return fmt.Errorf("fail: %w", err)
	}

	return nil
}

func renderOutbounds(conns map[string][]string) (rv string) {
	var b strings.Builder

	for k, v := range conns {
		b.WriteString(k)
		b.WriteString(": ")
		b.WriteString(strings.Join(v, "; "))
		b.WriteString("\r\n")
	}

	return b.String()
}
