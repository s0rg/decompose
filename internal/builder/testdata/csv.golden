service,listen,outbounds
1,"tcp:1
tcp:2","2: tcp:2
3: tcp:3
"
2,tcp:2,"1: tcp:1; tcp:2; tcp:3
3: tcp:3
"
3,tcp:3,
